### ISSUE TITLE
Groups item on the sharing modal

### ISSUE THREAD

**Issue by saleiva** [CartoDB/cartodb#4851]
@viddo edit: some planning

An user can:
- [x] see groups (that the user is member of) in sharing lists for both dataset and maps
  - [x] group name, 
  - [x] default avatar (since groups don't have avatars)
  - [x] show member count
    - [x] requires #5529 to return group users on grantables API too
- [x] search groups (by name) in sharing lists
- [x] grant access to whole organization, which should be reflected for all users
- [x] grant access to a group, which should be reflected on all users which are members of that group
- [x] not change inherited access, but see which ones are available (UI details in #5521)
  - [x] see which org/groups that the access is inherited from
  - [x] requires #5556 to return users groups on grantables' API 

Technicalities:
- use new `grantables` API endpoint to render a mixed list of users/groups, ordered by name alphabetically and ascending, see #5160
- use existing ACL structure/API endpoint for modifying access (now supports `group` type)

These use-cases have also been identified to be updated due to introducing groups, but since they're not blocking this issue or breaks they can postponed, to be solved in isolation:
- #5533 editor header, indicator for shared entities to account for groups
- #5534 change-privacy modal, shared-with info to account for groups
- #5575 delete vis (dataset/map), affected entities to take into account groups
- #5576 warn user if revoking access of a group that's using the visualization

**Comment by saleiva** [comment]
![dashboard - user private - datasets privacy share friends list](https://cloud.githubusercontent.com/assets/132149/9111718/f1fc2ebc-3c49-11e5-84b5-def40de6e9c5.png)

**Comment by saleiva** [comment]
Assigning to @viddo but pending on @juanignaciosl orders :)

**Comment by juanignaciosl** [comment]
Ok. @viddo, we'll begin groups api tomorrow, starting from the very bottom (PostgreSQL extension). I should get to this point next week.

**Comment by viddo** [comment]
Ok, I'd need some context information on what the expectations are here

**Comment by juanignaciosl** [comment]
In this milestone (Torrenueva) we're focused on backend and we're probably not beginning frontend work until next one, but @viddo and me have had a conversation about this and we wanted to write down share some details about the implementation.

Current design lists groups mixed with users, searchable and paginated server side. We can't query both endpoints and mix the results in the client because we want to support huge organizations, and retrieving and searching all users in the client is something we've already dismissed.

We're probably creating a new endpoint that allows querying for "entities I can share things with", and returning "ShareableWith" models that can be either users or groups (maybe also the organization), and probably modifying the current sharing endpoint to receive this kind of "virtual model" as well, since it makes no sense delegating to the frontend the logic of "is this a user or a group?".

cc @saleiva @javisantana @Kartones

**Comment by Kartones** [comment]
This had to come, @carlostallon wanted also user search in a near future so I guess the time is coming to do it.

And yes, I'd directly dismiss client-side filtering, etc., "as a backend engineer I'd prefer to have same functionality available to any API usage", makes more sense and is more powerful.

About "virtual models" I'm also ok, is what I attempted to do with Permissions when doing the multiuser, although maybe not ended so cool and only got used for tables.

**Comment by javisantana** [comment]
Could you create a google doc with the new endpoints and descriptions so we can discuss it there?

Do we need virtual models when we have two actual resources?

**Comment by Kartones** [comment]
I didn't went into details, but what I built for permissions was just `entityId`s so we could use anything that had a UUID (vis, tables). Then in code you implement the details, but in general sharing, searching and similar actions that have multiple targets should have an intermediate layer abstracting from the details (so the action is generic to all of them).

You can always have a `seach/users/rambo` and `search/groups/charlies` but the moment you need `search/mixed/whatever`...

**Comment by juanignaciosl** [comment]
https://docs.google.com/document/d/1SUoIGL-eFlMr0H-12Hr-MiJ-QtzVP1m6L7m_aCciaII/edit

**Comment by juanignaciosl** [comment]
@saleiva, take a look at that document, "Iterate on the design proposal" section. That's a @viddo proposal about implementing this feature. If you don't change the design proposal we'll do "a" alternative.

**Comment by saleiva** [comment]
I guess my answers are related to the a option

**Comment by viddo** [comment]
@juanignaciosl _show member count_ always shows "0 members" now, since the grantables API don't return users list, would it be possible to return always the users list of each group? or count as a metadata point at least?

**Comment by juanignaciosl** [comment]
@viddo ok. #5529.

**Comment by viddo** [comment]
closed by #5610
